core/genesis.go

```diff
@@ -176,67 +175,66 @@ func (e *GenesisMismatchError) Error() string {
 // specify a fork block below the local head block). In case of a conflict, the
 // error is a *params.ConfigCompatError and the new, unwritten config is returned.
 func SetupGenesisBlock(
-	db ethdb.Database, genesis *Genesis, lastAcceptedHash common.Hash, skipChainConfigCheckCompatible bool,
-) (*params.ChainConfig, error) {
+	db ethdb.Database, triedb *trie.Database, genesis *Genesis, lastAcceptedHash common.Hash, skipChainConfigCheckCompatible bool,
+) (*params.ChainConfig, common.Hash, error) {
 	if genesis == nil {
-		return nil, ErrNoGenesis
+		return nil, common.Hash{}, ErrNoGenesis
 	}
 	if genesis.Config == nil {
-		return nil, errGenesisNoConfig
+		return nil, common.Hash{}, errGenesisNoConfig
 	}
 	// Make sure genesis gas limit is consistent in SubnetEVM fork
-	gasLimitConfig := genesis.Config.FeeConfig.GasLimit.Uint64()
-	if gasLimitConfig != genesis.GasLimit {
-		return nil, fmt.Errorf("gas limit in fee config (%d) does not match gas limit in header (%d)", gasLimitConfig, genesis.GasLimit)
+	if genesis.Config.IsSubnetEVM(genesis.Timestamp) {
+		gasLimitConfig := genesis.Config.FeeConfig.GasLimit.Uint64()
+		if gasLimitConfig != genesis.GasLimit {
+			return nil, common.Hash{}, fmt.Errorf("gas limit in fee config (%d) does not match gas limit in header (%d)", gasLimitConfig, genesis.GasLimit)
+		}
+		// Verify config
+		if err := genesis.Config.Verify(); err != nil {
+			return nil, common.Hash{}, err
+		}
```

Comment:
Why are we moving this to inside of `IsSubnetEVM` ?

Reply:
Config verification is added with Subnet-EVM.

Reply:
We should always enforce this regardless of when SubnetEVM is enabled. If we don't, then we are no longer enforcing the invariant that chains are started with a verified set of precompile/state upgrades. Can we move this back out of the `IsSubnetEVM` branch?

Since it was enforced before, I don't see any reason to move it into this branch.

Reply:
Yes I agree with above, we should keep the fee check in IsSubnetEVM but not `Config.Verify`

Reply:
`SetupGenesisBlock` call path has changed. It was only used in eth/backend before which called only by VM. With these changes now it's being called by `NewBlockchain` which is heavily used by tests and `simulated`.  If we want to move them out of the subnet-evm then we need to make sure that all tests use a FeeConfig and proper gas limit.

Reply:
I suggest we do something like: https://github.com/ava-labs/subnet-evm/pull/749/commits/03560a4ad8d2bd751a3e82b44c3f913f91e82fe8

Reply:
Ah I see,  I didn't follow the reason for the change originally.

Since we add the need for calling `Verify()` from our own code. Would it make sense to call this from `vm.go` when initializing the VM? I think we can also move the FeeConfig gas limit matches the genesis config check into the same spot.

If we move it into `IsSubnetEVM`, then we will rely on not enforcing that, which will become an issue in future mandatory hard forks where we need to enable those and still don't want to enforce this config check.

Reply:
I think it makes sense